src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,184 @@
+{X Y : left_invariant_derivation I G} {f f' : C^∞⟮I, G; 𝕜⟯} {r : 𝕜}
+
+@[simp] lemma to_fun_eq_coe : X.to_fun = ⇑X := rfl
+@[simp] lemma coe_fn_coe : ⇑(X : C^∞⟮I, G; 𝕜⟯ →ₗ[𝕜] C^∞⟮I, G; 𝕜⟯) = X := rfl
+@[simp] lemma to_derivation_eq_coe : X.to_derivation = X := rfl
+
+lemma coe_injective (h : ⇑X = Y) : X = Y :=
+by { cases X, cases Y, congr', exact derivation.coe_injective h }
+
+@[ext] theorem ext (h : ∀ f, X f = Y f) : X = Y :=
+coe_injective $ funext h
+
+variables (X Y f)
+
+@[simp] lemma coe_derivation :
+  ⇑(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = (X : C^∞⟮I, G; 𝕜⟯ → C^∞⟮I, G; 𝕜⟯) := rfl
+
+lemma coe_derivation_injective : function.injective
+  (coe : left_invariant_derivation I G → derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) :=
+λ X Y h, by { cases X, cases Y, congr, exact h }
+
+/-- Premature version of the lemma. Prefer using `left_invariant` instead. -/
+lemma left_invariant' : (𝒅(𝑳 I g)) (1 : G) (derivation.eval_at (1 : G) ↑X) f =
+  derivation.eval_at g ↑X f := by rw [←to_derivation_eq_coe]; exact left_invariant'' X f g
+
+instance : has_zero (left_invariant_derivation I G) := ⟨⟨0, λ f g,
+  by simp only [linear_map.map_zero, derivation.coe_zero]⟩⟩
+
+instance : inhabited (left_invariant_derivation I G) := ⟨0⟩
+
+instance : has_add (left_invariant_derivation I G) :=
+{ add := λ X Y, ⟨X + Y, λ f g, by simp only [linear_map.map_add, derivation.coe_add,
+  left_invariant', pi.add_apply]⟩ }
+
+instance : has_neg (left_invariant_derivation I G) :=
+{ neg := λ X, ⟨-X, λ f g, by simp only [linear_map.map_neg, derivation.coe_neg, left_invariant',
+  pi.neg_apply]⟩ }
+
+instance : has_sub (left_invariant_derivation I G) :=
+{ sub := λ X Y, ⟨X - Y, λ f g, by simp only [linear_map.map_sub, derivation.coe_sub,
+  left_invariant', pi.sub_apply]⟩ }
+
+@[simp] lemma coe_add : ⇑(X + Y) = X + Y := rfl
+@[simp] lemma coe_zero : ⇑(0 : left_invariant_derivation I G) = 0 := rfl
+@[simp] lemma coe_neg : ⇑(-X) = -X := rfl
+@[simp] lemma coe_sub : ⇑(X - Y) = X - Y := rfl
+@[simp, norm_cast] lemma lift_add :
+  (↑(X + Y) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = X + Y := rfl
+@[simp, norm_cast] lemma lift_zero :
+  (↑(0 : left_invariant_derivation I G) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = 0 := rfl
+@[simp] lemma map_add : X (f + f') = X f + X f' := is_add_hom.map_add X f f'
+@[simp] lemma map_zero : X 0 = 0 := is_add_monoid_hom.map_zero X
+@[simp] lemma map_neg : X (-f) = -X f := linear_map.map_neg X f
+@[simp] lemma map_sub : X (f - f') = X f - X f' := linear_map.map_sub X f f'
+
+instance : add_comm_group (left_invariant_derivation I G) :=
+{ add_assoc := λ X Y Z, ext $ λ a, add_assoc _ _ _,
+  zero_add := λ X, ext $ λ a, zero_add _,
+  add_zero := λ X, ext $ λ a, add_zero _,
+  add_comm := λ X Y, ext $ λ a, add_comm _ _,
+  add_left_neg := λ X, ext $ λ a, add_left_neg _,
+  ..left_invariant_derivation.has_zero,
+  ..left_invariant_derivation.has_add,
+  ..left_invariant_derivation.has_neg }
```

Comment:
With the `injective` suggestion above:
```suggestion
coe_injective.add_comm_group _ coe_zero coe_add coe_neg coe_sub
```

Reply:
I am not able to do the same thing for module because lean does not guess the coercion, but it should be doable no?

Reply:
Yes. Definitely doable, I can work it out for your in a day or two if you can't work it out. Have a look for other uses of `injective.module` to work out the trick

Reply:
Ok I managed by looking at examples, although I did not find a shorter way to write the coercion to function that Lean recognizes.